expression/collation.go

```diff
@@ -168,131 +192,272 @@ func deriveCoercibilityForColumn(c *Column) Coercibility {
+		for _, c := range str {
+			if c >= 0x80 {
+				return false
+			}
+		}
+		return true
+	case charset.CharsetLatin1:
+		// for backward compatibility, we do not block SQL like select '啊' = convert('a' using latin1) collate latin1_bin;
+		return true
+	case charset.CharsetUTF8, charset.CharsetUTF8MB4:
+		// string in tidb is actually use utf8mb4 encoding.
+		return true
+	case charset.CharsetBinary:
+		// convert to binary is always safe.
+		return true
+	default:
+		e, _ := charset.Lookup(dstChs)
+		_, err := e.NewEncoder().String(str)
+		return err == nil
+	}
 }
 
 // inferCollation infers collation, charset, coercibility and check the legitimacy.
-func inferCollation(exprs ...Expression) (dstCollation, dstCharset string, coercibility Coercibility, legal bool) {
-	firstExplicitCollation := ""
-	coercibility = CoercibilityIgnorable
-	dstCharset, dstCollation = charset.GetDefaultCharsetAndCollate()
-	for _, arg := range exprs {
-		if arg.Coercibility() == CoercibilityExplicit {
-			if firstExplicitCollation == "" {
-				firstExplicitCollation = arg.GetType().Collate
-				coercibility, dstCollation, dstCharset = CoercibilityExplicit, arg.GetType().Collate, arg.GetType().Charset
-			} else if firstExplicitCollation != arg.GetType().Collate {
-				return "", "", CoercibilityIgnorable, false
+func inferCollation(exprs ...Expression) *ExprCollation {
+	if len(exprs) == 0 {
+		// TODO: see if any function with no arguments could run here.
+		dstCharset, dstCollation := charset.GetDefaultCharsetAndCollate()
+		return &ExprCollation{
+			Coer:      CoercibilityIgnorable,
+			Repe:      UNICODE,
+			Charset:   dstCharset,
+			Collation: dstCollation,
+		}
+	}
+
+	repertoire := exprs[0].Repertoire()
+	coercibility := exprs[0].Coercibility()
+	dstCharset, dstCollation := exprs[0].GetType().Charset, exprs[0].GetType().Collate
+	unknownCS := false
+
+	for _, arg := range exprs[1:] {
+		if dstCollation == charset.CollationBin || arg.GetType().Collate == charset.CollationBin {
+			if coercibility > arg.Coercibility() || (coercibility == arg.Coercibility() && arg.GetType().Collate == charset.CollationBin) {
+				coercibility, dstCharset, dstCollation = arg.Coercibility(), arg.GetType().Charset, arg.GetType().Collate
 			}
-		} else if arg.Coercibility() < coercibility {
-			coercibility, dstCollation, dstCharset = arg.Coercibility(), arg.GetType().Collate, arg.GetType().Charset
-		} else if arg.Coercibility() == coercibility && dstCollation != arg.GetType().Collate {
-			p1 := collationPriority[dstCollation]
-			p2 := collationPriority[arg.GetType().Collate]
-
-			// same priority means this two collation is incompatible, coercibility might derive to CoercibilityNone
-			if p1 == p2 {
-				coercibility, dstCollation, dstCharset = CoercibilityNone, getBinCollation(arg.GetType().Charset), arg.GetType().Charset
-			} else if p1 < p2 {
-				dstCollation, dstCharset = arg.GetType().Collate, arg.GetType().Charset
+			repertoire |= arg.Repertoire()
```

Comment:
Could we put this line below line378？ It seems that all cases use this logic except `coercibility == CoercibilityExplicit`.

Reply:
no, it will change repertoire value before use it.